properly render shortcut links w/ code formatting

aogier opened the pull request:
Hi, this is a tentative fix for #40. I've introduced some logic if we're working w/ state, and also had to move links' url inline otherwise they were broken if not at the end of line (see [tests](https://github.com/Byron/pulldown-cmark-to-cmark/compare/main...aogier:bugfix/40-shortcut_code?expand=1#diff-8ae4f4b9fc95a4103fc365a0a8631e4c9917171882848c967e23e61f900ff212R23)), let me know on that, maybe it could be made better in a next iteration on that.
Thank you, regards

Fixes #40

Byron commented on src/lib.rs:
```diff
@@ -275,6 +286,9 @@ where
+                if let Some(shortcut_text) = state.current_shortcut_text.as_mut() {
+                    shortcut_text.push_str(&format!("`{}`", text));
+                }
                 if let Some(text_for_header) = state.text_for_header.as_mut() {
```
Right now it's seemingly possible to use the same text twice, once for shortcuts and once for the header. I have a feeling this can't really be possible, so ideally that's represented in the type-system, or in the way the code behaves, at the least there should be an assertion to express this constraint.

Byron commented on src/lib.rs:
```diff
@@ -149,7 +149,11 @@ where
     F: fmt::Write,
 {
-    let mut state = state.unwrap_or_default();
+    let mut event_by_event = true;
```
Is this something that can be activated as part of the event handling and be made part of the state?

To me it seems strange that there is a special case based on the state being present or not, and I fear it works only because it's specifically made to make the 'event-by-event' test pass.

Leaving it off definitely leads to some unmet expectations in the test suite.

Byron reviewed (changes requested):
Thanks a lot for tackling this.

When reviewing the PR I was baffled due to my trouble to understand what 'event_by_event' is supposed to do until I realized that it specifically exists to add new logic without breaking the old one.

This leaves me with the question if there is a solution to this that doesn't need such a flag, or whether the current implementation is generally making that impossible.

Either way, with 'event_by_event' present this PR cannot be merged, and I have a feeling that the difficulties arise from generally somewhat incorrect link handling that requires more thought and refactoring to 'fall into place'.

Maybe for this one, it's necessary to take two steps back before being able to move forward.

aogier commented:
Hi, thank you for your feedback! After looking at #39, #40 and the current `main`, I've tested the current implementation against a "plain" and `STATE_TEST` stupicat and found a regression in the latter:
```markdown
# plain stupicat
$ cat <<EOF | cargo run --quiet --example=stupicat -- /dev/stdin; echo
> A [code], then some text.

[code]: https://blah
> EOF
A [code], then some text.

[code]: https://blah

# state test
$ export STUPICAT_STATE_TEST=1; cat <<EOF | cargo run --quiet --example=stupicat -- /dev/stdin; echo
> A [code], then some text.

[code]: https://blah
> EOF
A [code]

[code]: https://blah, then some text.
```
so now we have two bugs:
* code in links
* insufficient testing I've done in the last few contributions which led to improper link formatting when explicitly using `State`

This PR try to (naively) address either problems.

## code in links

The modification I've implemented satisfy tests I wrote, but maybe they're insufficient and I definitely not studied *a lot* the state machinery, I could start writing some more tests which fail perhaps and then start from here

## shortcut links where explicit `state` in `cmark*()`

It seems to me that independently from the implementation, correctly displaying shortcut links in a sensibly positioned footer is non-trivial/not possible. So I've modified current implementation so:
* when non-shortcut links or (shortcut links and explicit state in `cmark*()`) => link inlined
* when shortcut links and not explicit state in `cmark*()` => link in footer
This stem from the naive, empyrical observation that only when `state` is initialized from outside the cmark machinery we could face this need, hence the `event_by_event` marker: ~I'm a coward~ in order to stay unobtrusive I didn't fiddled with `state` itself. This is not a trick to cope w/ tests, but rather an ugly for coping with the scenario described above.  

Here are my cent and an half :)

Byron commented:
Thanks for elaborating, you definitely know more than me about this by now as I am somewhat in maintenance mode with this crate.

Having had a radical idea, I wonder if this issue would go away if there was no exposed `State`, hence all events have to be passed at once. Is this a feature anybody is using anyway?

If the answer is going against keeping `State` around, one could go down that route, too. Technically I still believe using State explicitly forces one to deal with it very properly, but if there is no actual user of it, it might just be a form of over-engineering which now shows its teeth.

What do you think?

aogier commented:
Hello, thank you for your prompt feedback!

Regarding `State`, while I personally never saw any use for it, at this moment it could be possible to support it with no big effort while we deprecate the relevant signatures so the people who indeed use it, if any, can have time to express their opinion. 

In the meantime I could perhaps despaghettify :stuck_out_tongue: a bit the current code!

What's your take on that? Regards

Byron commented:
It sounds like a great idea to go through a deprecation round first, which could also reveal if there are any users for it to help decide whether to go through with it. Simplification, however, isn't the only value here, it's whether or not this kind of link handling will benefit.

Thinking about it, the problem here is that if one passes events one by one, the current algorithm will print links after processing all (one) events, which puts them into the wrong spot. So the problem seems to be the test, rather than the state system itself: it's too unrealistic.

Now, if the method signature would contain information about completeness, or more specifically, whether or not to print the stored links, then everything should work as expected. Maybe one can provide an additional method signature that tells it whether or not the input events are 'complete'. That way `State` can definitely stay, while special handling through a flag can go.

aogier commented:
I think I should have worded my answer better: I am absolutely convinced that the mechanisms by which `state` is handled internally are well designed.  
I thought it might be a case of deprecating the exposure of `state` to the outside world as I think I understood was your idea in the previous comment (maybe I'm wrong as english isn't my first language yadda yadda :P)

That said, the part I'd like to desp^h^h^h^h _simplify_ is that nasty solution I found of using that `event_by_event` flag in case `state` is exposed/used from outside. Also, the idea of somehow signalling that a block is logically complete could also be a great addition!

Byron commented:
All good, all good, yadda yadda :), I am currently sketching an idea that I hope to push into this branch very soon. Maybe you can make sure the checkbox is ticked to simplify things. Will be back.

Byron commented:
And there it is.

Let me know what you think, and if you think it's usable and you are happy with the refactoring, this PR can be merged.

aogier commented:
Looks like a great refactor to me, that `finalize` addition could pave the way for workflows we still haven't think about like splitting a big input to smaller pages while retaining footer concept and who know what other. Feature parity in between the two scenarios is very good as this the dreaded `event_by_event` removal :D so I'd definitely merge it!

Thank you!

Byron commented:
Great to hear! Here is the new (_mildly breaking_) release: https://github.com/Byron/pulldown-cmark-to-cmark/releases/tag/v10.0.0
